### PR TITLE
added BarChart

### PR DESCRIPTION
**Pull request by mindas2000**

### CODE REVIEW

**Review by batiFinkelshtein — APPROVED** [review]
looks good

File: client-side/src/components/statistics.jsx [hunk 1]
```diff
@@ -1,30 +1,37 @@
...
 import VisitedWebsitesComponent from './statistics/graphs.jsx';
-
+import GraphBar from './statistics/graphBar.jsx';
+import RadioButtonComponent from '../stories/RadioButton/radio-Button.jsx';
 const Statistics = () => {
     const [startDate, setStartDate] = useState('');
     const [endDate, setEndDate] = useState('');
-    const [showVisitedWebsites, setShowVisitedWebsites] = useState(false);
+    const [show, setShow] = useState(false);
+    const [whatShow, setWhatShow] = useState('BarChart');
```
Comment (EA171) [inline client-side/src/components/statistics.jsx:11]: You shouldn't have 2 states, case of whatShow === undefined could be the same as show ==== false.
rename whatShow to selectedBar or something like that

File: client-side/src/components/statistics/graphBar.jsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import { useQuery } from '@apollo/client';
+import { BarChart } from '@mui/x-charts/BarChart';
+import { useAppSelector } from '../../redux/store.jsx';
+import Loader from '../../stories/loader/loader.jsx'
+import { getVisitedWebsitesByDate, formatDate } from './graphsUtils.js';
+import { GET_USERS } from './queries.js';
+
+export default function GraphBar({ startDate, endDate }) {
```
Comment (EA171) [inline client-side/src/components/statistics/graphBar.jsx:9]: add prop types and default values if not marked as isRequired

File: client-side/src/components/statistics/graphBar.jsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import { useQuery } from '@apollo/client';
+import { BarChart } from '@mui/x-charts/BarChart';
+import { useAppSelector } from '../../redux/store.jsx';
+import Loader from '../../stories/loader/loader.jsx'
+import { getVisitedWebsitesByDate, formatDate } from './graphsUtils.js';
+import { GET_USERS } from './queries.js';
+
+export default function GraphBar({ startDate, endDate }) {
+  const users = useQuery(GET_USERS);
+  const user = useAppSelector((state) => state.user.currentUser);
```
Comment (EA171) [inline client-side/src/components/statistics/graphBar.jsx:11]: why you can't take the current user from redux?

File: client-side/src/components/statistics/graphBar.jsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  startDate = formatDate(new Date(startDate.$d));
+  endDate = formatDate(new Date(endDate.$d));
```
Comment (EA171) [inline client-side/src/components/statistics/graphBar.jsx:13]: this rows are override the values that come from props, if you didn't need the props value, please remove it from there

File: client-side/src/components/statistics/graphBar.jsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  const result1 = getVisitedWebsitesByDate(users, user, startDate);
+  const result2 = getVisitedWebsitesByDate(users, user, endDate);
```
Comment (EA171) [inline client-side/src/components/statistics/graphBar.jsx:15]: please rename to more meaningful name

File: client-side/src/components/statistics/graphsUtils.js [hunk 6]
```diff
@@ -41,3 +41,23 @@ export function getWebsites(websites, users, user, dateStart, dateEnd) {
 
     return website;
 }
+
+export function getVisitedWebsitesByDate(users, user, date) {
+    const currentUser = users?.data?.users?.find(u => u.email === user.email);
```
Comment (EA171) [inline client-side/src/components/statistics/graphsUtils.js:46]: why do you need this search in users data if you send the currentuser also?
other option: think about taking the current user from the redux

File: client-side/src/components/statistics/graphsUtils.js [hunk 7]
```diff
@@ -41,3 +41,23 @@ export function getWebsites(websites, users, user, dateStart, dateEnd) {
...
     return website;
 }
+
+export function getVisitedWebsitesByDate(users, user, date) {
+    const currentUser = users?.data?.users?.find(u => u.email === user.email);
+    const visitedWeb = {};
+    currentUser?.visitsWebsites.forEach(visit => {
+        visit.visitsTime.forEach(visitTime => {
+            const visitDate = formatDate(new Date(visitTime.visitDate));
+            if (visitDate.split('-')[0] === date.split('-')[0] && parseInt(visitDate.split('-')[1], 10) === parseInt(date.split('-')[1], 10)) {
```
Comment (EA171) [inline client-side/src/components/statistics/graphsUtils.js:51]: maybe the formatDate function that you created will help you here?

**Review by EA171 — CHANGES_REQUESTED** [review]

File: client-side/src/components/statistics.jsx [hunk 8]
```diff
@@ -1,30 +1,37 @@
...
+        setWhatShow(event.target.value);
+    };
     const appolo_server_url = process.env.REACT_APP_APOLLO_SERVER_URL;
 
     const client = new ApolloClient({
         uri: appolo_server_url,
         cache: new InMemoryCache()
     });
-
+    const arr = [{ label: "PieChart", value: "PieChart" }, { label: "BarChart", value: "BarChart" }];
```
[the remark this answers is not in the document]
Comment (mindas2000) [reply client-side/src/components/statistics.jsx:28]: I change it

File: client-side/src/components/statistics.jsx (hunk 1 above)
Comment (mindas2000) [reply client-side/src/components/statistics.jsx:11]: I did it, thanks

File: client-side/src/components/statistics/graphBar.jsx (hunk 2 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:9]: I added

File: client-side/src/components/statistics/graphBar.jsx (hunk 4 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:13]: I need the accessories, I just used the same variable, is it better if I use a new variable?

File: client-side/src/components/statistics/graphBar.jsx (hunk 5 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:15]: I changed it to dataOfEndDate and dataOfStartDate.

File: client-side/src/components/statistics/graphsUtils.js (hunk 7 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphsUtils.js:51]: I fixed it.

File: client-side/src/components/statistics/graphBar.jsx (hunk 3 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:11]: I took from redux, it isn't right?

Comment (EA171) [reply client-side/src/components/statistics/graphBar.jsx:12]: so why do you need to query all the users?

File: client-side/src/components/statistics/graphBar.jsx (hunk 4 above)
Comment (EA171) [reply client-side/src/components/statistics/graphBar.jsx]: yes, it will be more clear if you will assign it to a new variables something like: formattedStartDate

File: client-side/src/components/statistics/graphsUtils.js (hunk 6 above)
Comment (EA171) [reply client-side/src/components/statistics/graphsUtils.js:46]: so please go to the peace of code that insert the user to the redux and check there which data is saved to the redux. if you are missing data there, add it to the redux

File: client-side/src/components/statistics/graphBar.jsx (hunk 4 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx]: I change it.

File: client-side/src/components/statistics/graphBar.jsx (hunk 3 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:12]: So do I need to change something?

Comment (EA171) [reply client-side/src/components/statistics/graphBar.jsx:12]: I hope you will success to take the user from the redux so this useQuery for getting the users will be unnecessary you will be able to remove it

File: client-side/src/components/statistics/graphsUtils.js (hunk 6 above)
Comment (EA171) [reply client-side/src/components/statistics/graphsUtils.js:46]: see what I write you in slack

File: client-side/src/components/statistics/graphBar.jsx (hunk 3 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphBar.jsx:12]: I can remove it and use it from Redux, but I don't understand why the apollo-server that I worked on for so long is needed, it is also possible in graphs not to use it but from Redux

File: client-side/src/components/statistics/graphsUtils.js (hunk 6 above)
Comment (mindas2000) [reply client-side/src/components/statistics/graphsUtils.js]: I answered you.

**Review by EA171 — CHANGES_REQUESTED** [review]

File: server-side/controllers/user.controller.js [hunk 9]
```diff
@@ -1,34 +1,33 @@
...
-  const {id} = req.params;
-  if(!mongoose.Types.ObjectId.isValid(id))
-    return next({message:'id is not valid'})
+export const getUserById = async (req, res, next) => {
+  const { id } = req.params;
+  if (!mongoose.Types.ObjectId.isValid(id))
+    return next({ message: 'id is not valid' })
   try {
-    const user = await getUserById_service(id);    
+    const user = await Users.findById(id).populate('visitsWebsites profiles preferences').select('-__v');
```
Comment (EA171) [inline server-side/controllers/user.controller.js:23]: access to collections and logics should be in service file. why did you remove it?

File: server-side/controllers/user.controller.js [hunk 10]
```diff
@@ -88,4 +87,16 @@ export const updatedUser = async (req, res,next) => {
   }
 };
 
-
+export const getUserByEmail = async (req, res, next) => {
+  const email = req.params.email;
```
Comment (EA171) [inline server-side/controllers/user.controller.js:91]: use destructure objects: const {email} = req.params

File: server-side/controllers/user.controller.js [hunk 11]
```diff
@@ -88,4 +87,16 @@ export const updatedUser = async (req, res,next) => {
...
+  try {
+    const user = await Users.findOne({ email }).populate({ path: 'visitsWebsites', populate: { path: 'websiteId' } }).populate('profiles preferences').select('-__v');
```
Comment (EA171) [inline server-side/controllers/user.controller.js:93]: access to collections and logics should be in service file.

File: server-side/controllers/user.controller.js [hunk 12]
```diff
@@ -88,4 +87,16 @@ export const updatedUser = async (req, res,next) => {
...
+    if (!user) {
+      return next({ message: 'user not found ', status: 404 })
+    }
+    res.send(user);
```
Comment (EA171) [inline server-side/controllers/user.controller.js:97]: add return statement and status code

File: server-side/controllers/user.controller.js [hunk 13]
```diff
@@ -88,4 +87,16 @@ export const updatedUser = async (req, res,next) => {
...
+  } catch (err) {
+    console.error(err);
+    next({ message: err.message, status: 500 })
```
Comment (EA171) [inline server-side/controllers/user.controller.js:100]: add return statement

**Review by EA171 — CHANGES_REQUESTED** [review]

File: server-side/controllers/user.controller.js (hunk 9 above)
Comment (mindas2000) [reply server-side/controllers/user.controller.js:23]: I changed it.

File: server-side/controllers/user.controller.js (hunk 10 above)
Comment (mindas2000) [reply server-side/controllers/user.controller.js:91]: ✅

File: server-side/controllers/user.controller.js (hunk 11 above)
Comment (mindas2000) [reply server-side/controllers/user.controller.js:93]: I changed it.

File: server-side/controllers/user.controller.js (hunk 13 above)
Comment (mindas2000) [reply server-side/controllers/user.controller.js:100]: ✅

**Review by EA171 — APPROVED** [review]
